src/robot/running/keywordrunner.py

```
@@ -206,16 +249,69 @@ def _to_number_with_arithmetics(self, item):
+            if not is_list_like(item):
+                # Checking here because _validate() doesn't have access
+                # to the actual values.
+                raise DataError("For-In-Zip Loop items must all be List-like (but not Strings); got %s with value '%s'" % (item.__class__, item))
+        for zipped_item in zip(*items):
+            answer.extend(zipped_item)
+        return answer
+
+class ForInEnumerateLoopRunner(BasicForLoopRunner):
+    def __init__(self, forstep, templated=False):
+        super(ForInEnumerateLoopRunner, self).__init__(forstep, templated)
+
+    def _flavor_name(self):
+        return 'IN ENUMERATE'
+
+    def _transform_items(self, items):
+        answer = list()
+        for zipped_item in zip(*items):
+            answer.extend(zipped_item)
+        return answer
+
+    @property
+    def _original_values_per_iteration(self, values):
+        """
+        The number of values per iteration;
+        used to check if we have (a multiple of this) values.
+
+        This is its own method to support loops like ForInEnumerate
+        which add/remove items to the pool.
+        """
+        if len(values) < 2:
+            raise DataError('FOR IN ENUMERATE expected 2+ loop variables, got %d.'
+                    % len(values))
+        return len(values) - 1
+
+    def _get_values_for_one_round(self, data):
+        for idx, values in enumerate(super(ForInEnumerateLoopRunner, self)
+                ._get_values_for_one_round(data)):
+            yield (idx,) + values
+
+
+# Defining this dict here in for potentially easier extension by libraries...
+# ...unless that's a terrible idea?
+_FOR_LOOP_FLAVORS = dict(
+        INRANGE=ForInRangeLoopRunner,
+        INZIP=ForInZipLoopRunner,
+        INENUMERATE=ForInEnumerateLoopRunner,
+        )
+def ForLoopRunner(context, templated=False, flavor='IN'):
+    flavor_key = flavor.replace(' ', '')
+    runner_class = _FOR_LOOP_FLAVORS.get(flavor_key, BasicForLoopRunner)
```

Hmm, wouldn't it be better to just fail with a `DataError` if the given flavor is invalid? That's something that definitely needs to be tested too.